x509/revocation/crl/crl.go

```diff
@@ -0,0 +1,143 @@
+)
+
+func (r RevocationReasonCode) String() string {
+	switch r {
+	case Unspecified:
+		return "Unspecified"
+	case KeyCompromise:
+		return "Key Compromise"
+	case CACompromise:
+		return "CA Compromise"
+	case AffiliationChanged:
+		return "Affiliation Changed"
+	case Superseded:
+		return "Superseded"
+	case CessationOfOperation:
+		return "Cessation Of Operation"
+	case CertificateHold:
+		return "Certificate Hold"
+		// STATUS CODE 7 IS NOT USED
+	case RemoveFromCRL:
+		return "Remove From CRL"
+	case PrivilegeWithdrawn:
+		return "PrivilegeWithdrawn"
+	case AACompromise:
+		return "AA Compromise"
+	default:
+		return "Unknown revocation reason code: " + strconv.Itoa(int(r))
+	}
+}
+
+var (
+	crlNumberExtensionOID        = asn1.ObjectIdentifier{2, 5, 29, 20}
+	revocationReasonExtensionOID = asn1.ObjectIdentifier{2, 5, 29, 21}
+	invalidityDateExtensionOID   = asn1.ObjectIdentifier{2, 5, 29, 24}
+)
+
+type crlNumberExtension struct {
+	ID        asn1.ObjectIdentifier
+	Critical  bool `asn1:"optional"`
+	CRLNumber int
+}
+
+// TODO: handle additional CRL Extensions
+// 2.5.29.21,cRLReason
+// id-ce-cRLReasons OBJECT IDENTIFIER ::= { id-ce 21 }
+// -- reasonCode ::= { CRLReasonCode }
+
+//2.5.29.24,invalidityDate
+// id-ce-invalidityDate OBJECT IDENTIFIER ::= { id-ce 24 }
+// -- InvalidityDate ::=  GeneralizedTime
+
+// RevocationData - information on whether a certificate has been
+// revoked by a specified CRL, and information on the CRL
+type RevocationData struct {
+	CRLSignatureAlgorithm         x509.SignatureAlgorithm
+	CRLSignatureValue             []byte
+	Version                       int `asn1:"optional,default:0"`
+	Issuer                        pkix.Name
+	ThisUpdate                    time.Time
+	NextUpdate                    time.Time `asn1:"optional"`
+	CRLExtensions                 ListExtensionData
+	UnknownCRLExtensions          []pkix.Extension `asn1:"tag:0,optional,explicit"`
+	IsRevoked                     bool
+	RevocationTime                time.Time
+	CertificateEntryExtensions    RevokedCertExtensionData
+	RawCertificateEntryExtensions []pkix.Extension `asn1:"optional"`
+}
+
+// ListExtensionData - Data from optional, non-critical pkix.CertificateList extensions
+type ListExtensionData struct {
+	CRLNumber int
+	AuthKeyID x509.SubjAuthKeyId `json:"authority_key_id,omitempty"`
+}
+
+// RevokedCertExtensionData - Data from optional, non-critical pkix.RevokedCertificate extensions
+type RevokedCertExtensionData struct {
+	Reason         string `json:"revocation_reason,omitempty"`
+	invalidityDate time.Time
+}
+
+func gatherListExtensionInfo(certList *pkix.CertificateList, ret *RevocationData) {
+	for _, extension := range certList.TBSCertList.Extensions {
+		if extension.Id.Equal(crlNumberExtensionOID) {
+			var ext crlNumberExtension
+			asn1.Unmarshal(extension.Value, &ext.CRLNumber)
+			ret.CRLExtensions.CRLNumber = ext.CRLNumber
+		} else {
+			ret.UnknownCRLExtensions = append(ret.UnknownCRLExtensions, extension)
```

Comment:
Would it make sense to separately track unknown *critical* extensions?

Reply:
That's a good idea. Added this with https://github.com/zmap/zcrypto/pull/142/commits/d367ce200be69c6c8990436ba81fdeb7237344d4